File: Dataset/Leetcode/train/5/52.cpp
class Solution
{
public:
    string XXX(string s)
    {
        int len = s.size();
        for (int trylen = len; trylen > 0; trylen--)
        {
            for (int i = 0; i + trylen <= len; i++)
            {
                if (check(s, i, i + trylen - 1))
                    return s.substr(i, trylen);
                else
                    continue;
            }
        }
        return "";
    }

    bool check(string &s, int l, int r)
    {
        if (r - l > 1)
        {
            if (s[l] == s[r])
                return check(s, l + 1, r - 1);
            else
                return false;
        }
        else if (r - l == 1)
        {
            if (s[l] == s[r])
                return true;
            else
                return false;
        }
        else
            return true;
    }
};

File: Dataset/Leetcode/train/20/52.java
 class Solution {
    public boolean XXX(String s) {
        LinkedList<Character> list = new LinkedList<>();
        char[] chars = s.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            if (isLeft(c)) {
                //左括号，入栈
                list.push(c);
            } else {
                //右括号，上一个左括号与之不匹配，不用再判断了，直接返回
                if (list.isEmpty() || !isMatch(list.pop(), c)) {
                    return false;
                }
            }
        }
        return list.isEmpty();
    }
    private boolean isLeft(char c) {
        return c == '(' || c == '[' || c == '{';
    }

    private boolean isMatch(char a, char b) {
        switch (a) {
            case '(':
                return b == ')';
            case '[':
                return b == ']';
            case '{':
                return b == '}';
            default:
                return false;
        }
    }
}

